growthbook/growthbook
# 390 (2) - Data Source editing UI/UX -> Identifier types

### Features and Changes

Adds delete functionality with a confirmation using existing delete button functionality (with some modifications to improve whitespace around the icon).

Some UI improvements:

- empty state
- `(no description)` in muted text when there's no description – previously this would look unbalanced if there was no description
- modal for adding and editing records
- makes the `userIdType` on the record not editable since changing it can break things. This forces users to go through the delete flow and confirm it's what they want to do.


Tickets:

- Related to #390 


### Dependencies

None

### Testing

- Perform Edit and Delete operations via the context menu
- Add via the `+ Add` buttons in both the empty state or the Identifier Types heading
- See screenshots below 

### Screenshots

![data source - user identifier](https://user-images.githubusercontent.com/113377031/191623384-2c297785-0fb4-4d19-8e4f-48890c0632c4.gif)


#### View

1. Add button pops open the modal to add a new item
2. Kebab-style context menu
3. Edit and Delete actions

![CleanShot 2022-09-21 at 15 09 51@2x](https://user-images.githubusercontent.com/113377031/191619519-94e38d86-235d-44e7-b756-482ee3b20b8c.png)


#### Edit

1. Modal title says "Edit Identifier Type"
2. Identifier Type field is read only

![CleanShot 2022-09-21 at 15 13 10@2x](https://user-images.githubusercontent.com/113377031/191619866-d1a33793-fb86-49b4-899a-2d1ebe00f85a.png)


#### Add

1. Modal title says "Add Identifier Type"
2. Identifier Type field is editable

![CleanShot 2022-09-21 at 15 14 34@2x](https://user-images.githubusercontent.com/113377031/191619989-c0300370-67b0-40dd-a3ae-c61d78c49f8a.png)


##### Validation

When adding a new record, we check to make sure the following criteria is met:

- the `userIdType` is not empty
- the item is not already in the `userIdTypes` (uniqueness constraint on `userIdType`)

![CleanShot 2022-09-21 at 16 42 14@2x](https://user-images.githubusercontent.com/113377031/191628759-e3cd09ae-c3a6-41d3-87d1-68af565301d1.png)

![CleanShot 2022-09-21 at 16 42 57@2x](https://user-images.githubusercontent.com/113377031/191628833-4ef07ca7-2e69-4a03-a93c-c5a6f41a3547.png)


#### Delete

Uses existing `DeleteButton` (with modifications to support more custom styling).

<img width="797" alt="image" src="https://user-images.githubusercontent.com/113377031/191620025-5c6bd20a-3dcf-495c-8c01-3ff3b4b3b075.png">


#### Empty state

Adds a card style empty state with primary CTA "+ Add"

<img width="1235" alt="image" src="https://user-images.githubusercontent.com/113377031/191621999-f26c0d75-9b1c-4b23-a1c1-b7fbe1629141.png">
